### PR TITLE
Add GreedyLR adaptive learning rate scheduler

### PR DESCRIPTION
**Pull request by balak4**
## Summary

- Add GreedyLR, a metric-based adaptive learning rate scheduler that adjusts the learning rate during training based on the current loss
- Based on ["Dynamic Learning Rate Scheduling based on Loss Changes Leads to Faster Convergence"](https://arxiv.org/abs/2512.14527)
- Integrated with Trainer following the existing ReduceLROnPlateau pattern

## Changes

- `optimization.py`: Add `GreedyLR` class, `StreamingAverage` helper, and `get_greedy_schedule()` factory
- `trainer.py`: Add GreedyLR alongside ReduceLROnPlateau for metric-based scheduler stepping
- `training_args.py`: Add validation requiring eval strategy for greedy scheduler
- `trainer_utils.py`: Add `GREEDY` to `SchedulerType` enum
- `__init__.py`: Export `GreedyLR` and `get_greedy_schedule`
- `tests/optimization/test_greedy_lr.py`: Comprehensive test suite (30 tests)
- `docs/source/en/main_classes/optimizer_schedules.md`: Autodoc entries
- `examples/greedy-lr/`: Example script and documentation

## How it works

GreedyLR adaptively adjusts the learning rate during training based on the current loss:
- **Increases** LR (divides by factor) when loss improves beyond patience
- **Decreases** LR (multiplies by factor) when loss stops improving beyond patience
- Supports warmup/cooldown periods, metric smoothing via streaming average, and automatic reset when LR hits min

## Usage

```python
training_args = TrainingArguments(
    lr_scheduler_type="greedy",
    lr_scheduler_kwargs={"patience": 10, "factor": 0.95, "min_lr": 1e-5},
    eval_strategy="steps",
    eval_steps=200,
    ...
)
```

## Test plan

- [x] `pytest tests/optimization/test_greedy_lr.py` — 30/30 pass
- [x] `pytest tests/optimization/test_optimization.py` — 4/4 pass (no regressions)
- [x] `make check-repo` — no new errors introduced
- [x] `from transformers import GreedyLR, get_greedy_schedule` works

### CODE REVIEW

**Review by redpanda1995 — APPROVED** [review]

**Comment by balak4** [comment]
@Rocketknight1 could you provide specific feedback on what led to closing this? I'd like to address any concerns.

For context: I'm one of the authors of this paper from our team at AWS (https://arxiv.org/abs/2512.14527). GreedyLR is a new adaptive LR scheduler that adjusts the learning rate during training based on the current loss. We validated it across NLP, CV, and LLM tasks up to 7B parameters, showing faster convergence without performance loss for both fine-tuning and pre-training.

The implementation follows the existing ReduceLROnPlateau integration pattern, includes 30 unit tests, and has been E2E tested on single and multi-GPU setups on AWS. All CI checks pass.

Happy to address any specific feedback or make changes to better fit the library's conventions.

cc: @w601sxs @pranavvm26 (co-authors)

**Comment by Rocketknight1** [comment]
Hi @balak4, I'm sorry! We've been getting swamped with code agent PRs lately, and so we've been trying to filter out code agent PRs from first-time contributors because a lot of them are low-quality spam and our reviewer time is too limited to read them all in detail. 

In this case the PR was more legitimate than I realized at first - I'll reopen it. Sorry for the inconvenience!

**Comment by balak4** [comment]
> Hi @balak4, I'm sorry! We've been getting swamped with code agent PRs lately, and so we've been trying to filter out code agent PRs from first-time contributors because a lot of them are low-quality spam and our reviewer time is too limited to read them all in detail.
> 
> In this case the PR was more legitimate than I realized at first - I'll reopen it. Sorry for the inconvenience!

@Rocketknight1 Gotcha, I understand, thanks for re-opening. Please let me know if you have any questions / need clarification. We're eager to share this work with the community!

**Comment by balak4** [comment]
Hey @Rocketknight1 , checking in since it's been 2 weeks since we submitted this PR - any update on when the PR could be reviewed?

**Comment by Rocketknight1** [comment]
Ah, sorry! I should have pinged a reviewer when I reopened it. cc @sunmarc!

File: examples/scheduler/README.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
+# GreedyLR: Adaptive Learning Rate Scheduler
+
+GreedyLR monitors training metrics and adaptively adjusts the learning rate -- increasing when improving, decreasing when plateauing. It works for both pre-training and fine-tuning.
+
+## Paper
+
+> **GreedyLR: A Novel Adaptive Learning Rate Scheduler**
+>
+> Despite significant advances in optimizers for training, most research works use common scheduler choices like Cosine or exponential decay. In this paper, we study GreedyLR, a novel scheduler that adaptively adjusts the learning rate during training based on the current loss. To validate the effectiveness of our proposed scheduler, we conduct experiments on several NLP, CV, and LLM tasks with up to 7B parameters, including both fine-tuning and pretraining experiments. The results show that our approach outperforms several state-of-the-art schedulers in terms of accuracy, speed, and convergence.
+
```
Comment (SunMarc) [inline examples/scheduler/README.md:10]: maybe create a new folder called examples/scheduler and put that there ? It will be better ! Also, maybe it will be interesting to have a section for scheduler in trainer docs cc @stevhliu

File: examples/greedy-lr/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+CLAUDE.md
+archive/
```
Comment (SunMarc) [inline examples/greedy-lr/.gitignore]: remove

File: tests/optimization/test_greedy_lr.py [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from transformers import is_torch_available
+from transformers.testing_utils import require_torch
+
```
Comment (SunMarc) [inline tests/optimization/test_greedy_lr.py:19]: add some trainer tests also to make sure the integration is correctly done. You need to update the` test_trainer_optimizers.py` file

**Review by SunMarc — COMMENTED** [review]
Thanks, just a few ntis

**Comment by balak4** [comment]
> Thanks, just a few ntis

@SunMarc thanks for the feedback! Will get these addressed

**Comment by balak4** [comment]
@SunMarc comments addressed. Please review and let me know if there's anything else you'd like updated. 

cc: @Rocketknight1

**Comment by balak4** [comment]
@SunMarc @Rocketknight1 hope you're well, just a friendly follow-up. Are we good to get this PR merged now that the comments are addressed?

**Review by SunMarc — APPROVED** [review]
Thanks a lot ! Sorry for the delay
